rfd/0090-db-mfa-sessions.md

```diff
@@ -0,0 +1,283 @@
+and a user may be prompt once per minute for their MFA.
+
+We should remove the 1 minute restriction on database cert TTL to improve UX.
+
+## Details
+When per-session-MFA is enabled, we should not restrict database cert TTL to 1 minute.
+
+Instead, database cert TTL should be restricted to `max_session_ttl`, and the cert
+should be kept in-memory by a local proxy tunnel.
+
+The commands `tsh db login` and `tsh proxy db` (without --tunnel) should return an error message if used when
+per-session-mfa is required, since these commands both save certs to disk.
+A user should be directed by the error message to use `tsh db connect` or `tsh proxy db --tunnel` instead:
+
+```bash
+$ tsh db login example-db
+ERROR: per-session-mfa is required to access database "example-db", but per-session-mfa is not supported
+by this command. Use `tsh db connect` or `tsh proxy db --tunnel` instead.
+See: $docs_reference for more details.
+
+$ tsh proxy db example-db
+ERROR: per-session-mfa is required to access database "example-db", but per-session-mfa is not supported
+by this command. Use `tsh db connect` or `tsh proxy db --tunnel` instead.
+See: $docs_reference for more details.
+```
+
+"Doesn't this just disable per-session-mfa for database access?" (My initial thinking)
+
+- No, not quite. Sessions are still limited by the lifetime of a local proxy process and an MFA prompt is always required to start these proxies when per-session-mfa is enabled.
+- This can "weaken" per-session-mfa security though. If a user finds that they would prefer the old behavior of 1-minute TTL certs, they can still achieve that through configuration using `max_session_ttl`.
+
+### API
+
+We should change the auth service to always issue database certs without capping the cert TTL to
+1 minute for per-session-MFA; clients will only keep the newly issued database certs in-memory
+when per-session-mfa is required.
+
+### Local Proxy With Tunnel
+
+Make the local proxy tunnel keep its database certs in-memory only, and to always request for cert reissue when the local proxy starts.
+This way, the certs are ephemeral and their lifetime is narrowed by the lifetime of the proxy.
+
+Right now, we load certs from the filesystem when using `tsh proxy db --tunnel`.
+So this change will affect per-session-MFA UX flow in that `tsh proxy db --tunnel` will always prompt for MFA when the local proxy starts.
+
+The local proxy can also refresh its certs as needed by invoking a callback function on each new connection which checks the local proxy's db cert expiry.
+
+### Local Proxy Without Tunnel
+
+If per-session-MFA is required, then we should forbid `tsh proxy db <database>` and give the user an error message telling them to add the `--tunnel` flag since per-session-MFA is required.
+
+This prevents per-session-MFA database certs from being saved to the filesystem and directs users towards the better UX of using a local proxy tunnel.
```

Comment:
I hadn't considered that. We may need to add a field somewhere telling the auth service whether the client is a proxy tunnel. I'll have to look into this, thanks for pointing it out

Reply:
FWIW a solution like that could point towards my other comment - changing the way we enforce per-session MFA for proxy tunnel connections.

Reply:
@smallinsky We discussed ignoring this issue, but before I mark this resolved I just want to be sure that you agree.  If we want to address the issue, then we would need to add the version of `tsh` sent in `UserCertsRequest` (I think, unless there is some other way to determine the version of `tsh` the client is using).

Reply:
@Joerger revisiting this - instead of making`tsh db login/connect` return an error when `per_session_mfa` is enabled, we can just enforce the 1 minute TTL (and print a warning/suggestion to use `tsh proxy db --tunnel` instead for long lived sessions?). The proxy tunnel can have a new cert request parameter like `DisablePerSessionMFATTL = true` and then all proxy tunnels going forward from v12.x will retain long-lived database certs in memory, while all other use-cases will have the 1 minute TTL applied. What do you think?

Reply:
@smallinsky @r0mant @zmb3  your thoughts as well please. I will need to amend the RFD draft if this is what we'd like to do to address older tsh clients. I unfortunately did not have time to get this done in v12 and hadn't revisited this RFD until now.

With this approach we don't disable any existing login flows (as originally proposed) and still get the benefit that GUI clients will not cause MFA prompt spam.

Reply:
I think that we have discussed  this but I can't find any reference. 
The the question is why why should prevent the older clients  to use previous behaviour ? The user can still alter the tsh code and save the db cert on disc from memory also any mechanism based on user param can by bypassed.

Reply:
My point here is that this new flow makes it possible to issue MFA-verified certificates with no TTL (or a longer TTL based on `max_session_ttl`). These certificates can be used with old/modified clients to bypass the expected MFA checks of DB access in general. They can also be potentially stolen by attackers to provide MFA-verified access past the 1-minute TTL, potentially with no TTL? 

The only way to prevent this downgrade in security is to have the Auth server enforce it. Right now, this is done by issuing MFA certificates with a 1-minute TTL. I'm not sure I have a good solution for replacing the 1-minute TTL...maybe we could add a grace period specifically for DB access through tunnel to allow recently expired MFA certificates, so that this doesn't affect other flows? Something server-enforced would be safer, so a client can't dictate the TTL of MFA certs themselves.

In my opinion, we should keep the default settings the most secure, so removing the 1-minute TTL by default seems a bit scary to me. My recommendation with the current design would be to add a new flag to configure this TTL, like `require_session_mfa_ttl` or `require_db_session_mfa_ttl`.

I just wanted to point out this change given that Per-session MFA is a security-focused features. If we think this isn't a significant security downgrade or is too theoretical, we can resolve this issue. Curious what others think.

Reply:
@Joerger totally understand the concern. TBH I have been nervous about actually implementing this feature for the same reasons (even though we've already approved the RFD).

If we let the client disable the 1-minute TTL, then only newer clients will get certs without the 1 minute TTL. Additionally, the newer clients will only disable that 1-minute TTL inside of the alpn local proxy tunnel, which holds the certs in memory only. So no version of tsh should save these longer-lived MFA certs to disk.

Now that being said, I have asked myself how customers would feel about this. Technically, a user in their cluster could build their own modified tsh that saves the certs to disk. Why they would want to do that - I don't know. But it would be possible. I can't see a malicious motive for doing that, as they already have the MFA token and login info to use their own account in the cluster. Perhaps they have some "innocent" reason for doing that - in which case the security posture of the cluster is weakened because users could "innocently" start saving certs to disk with their own modified tsh, and those certs could be exfiltrated by an attacker.

I think if that's a security downgrade that a customer finds unacceptable, then we could make this feature opt-in with configuration settings. For example:
```yaml
spec:
  require_session_mfa: true
  # defaults to 0h1m0s if not specified
  session_mfa_ttl: 0h5m0s # limit per-session MFA cert TTL to 5 minutes.
  # controls enforcement of per-session MFA cert TTL limit.
  # options are: "client", "on", "off"
  #   - client: allows clients to specify if requested MFA certs should be limited to `session_mfa_ttl`
  #   - on: enforce `session_mfa_ttl` max cert expiry always for per-session mfa certs.
  #   - off: disable enforcement of `session_mfa_ttl`
  session_mfa_ttl_enforcement: "client"
```

Benefit of adding configuration: allow customers to choose settings that fit their threat model.
Drawback: Even more configuration complexity (negative impact for time-to-value)

If we like the extra-configuration path, I'm open to suggestions around naming things 😅

Reply:
and of course we can get even more specific with the configuration, based on the type of session (db/app/ssh/etc). I'm only wary of adding further complexity to Teleport config.

Reply:
> They can also be potentially stolen by attackers to provide MFA-verified access past the 1-minute TTL, potentially with no TTL?

Regarding "potentially no TTL?": `max_session_ttl` is always enforced, and we have a maximum configurable limit on that, IIRC 30 hours. So Teleport will never issue any kind of cert without an expiration.

Reply:
I also think it's relevant to this discussion to add that we recently changed the interaction between `disconnect_expired_certs` and `require_session_mfa` TTL limits: https://github.com/gravitational/teleport/pull/18607

If `disconnect_expired_certs` is enabled, we now only disconnect certs based on the `tsh login` cert TTL rather than the 1-minute MFA session TTL. This (partially) helps the issue with DB GUI clients.

Reply:
@GavinFrazar @Joerger @smallinsky This was discussed on the product sync quite a long time ago and we agreed upon treating a single Teleport Connect session (which effectively equates to "tsh proxy db") as MFA session to fix GUI clients UX. We'll need to extend this approach to Kube Access as well in Q1.